cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/AvroFileManager.java

```diff
@@ -0,0 +1,300 @@
+    private final Location location;
+    private FSDataOutputStream outputStream;
+    private DataFileWriter<GenericRecord> dataFileWriter;
+    private long createTime;
+    private boolean isOpen = false;
+
+    public AvroFile(Location location) {
+      this.location = location;
+    }
+
+    /**
+     * Opens the underlying file for writing.
+     * If open throws an exception then underlying file may still need to be deleted.
+     *
+     * @throws IOException
+     */
+    void open() throws IOException {
+      try {
+        this.outputStream = new FSDataOutputStream(location.getOutputStream(), null);
+        this.dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<GenericRecord>(schema));
+        this.dataFileWriter.create(schema, this.outputStream);
+        this.dataFileWriter.setSyncInterval(syncIntervalBytes);
+        this.createTime = System.currentTimeMillis();
+        // Sync the file as soon as it is created, otherwise a zero length Avro file can get created on OOM
+        sync();
+      } catch (Exception e) {
+        close();
+        throw new IOException("Exception while creating file " + location, e);
+      }
+      this.isOpen = true;
+    }
+
+    public boolean isOpen() {
+      return isOpen;
+    }
+
+    public Location getLocation() {
+      return location;
+    }
+
+    public void append(ILoggingEvent event) throws IOException {
+      try {
+        dataFileWriter.append(co.cask.cdap.logging.serialize.LoggingEvent.encode(schema, event));
+      } catch (Exception e) {
+        close();
+        throw new IOException("Exception while appending to file " + location, e);
+      }
+    }
+
+    public long getPos() throws IOException {
+      try {
+        return outputStream.getPos();
+      } catch (Exception e) {
+        close();
+        throw new IOException("Exception while getting position of file " + location, e);
+      }
+    }
+
+    public long getCreateTime() {
+      return createTime;
+    }
+
+    public void flush() throws IOException {
+      try {
+        dataFileWriter.flush();
+        outputStream.hsync();
+      } catch (Exception e) {
+        close();
+        throw new IOException("Exception while flushing file " + location, e);
+      }
+    }
+
+    public void sync() throws IOException {
+      try {
+        dataFileWriter.flush();
+        outputStream.hsync();
+      } catch (Exception e) {
+        close();
+        throw new IOException("Exception while syncing file " + location, e);
+      }
+    }
+
+    @Override
+    public void close() throws IOException {
+      if (!isOpen) {
+        return;
+      }
+
+      LOG.trace("Closing file {}", location);
+      isOpen = false;
+
+      try {
+        if (dataFileWriter != null) {
+          dataFileWriter.close();
+        }
+      } finally {
+        if (outputStream != null) {
+          outputStream.close();
+        }
+      }
+    }
+  }
+
+  class LoggingLocationFactory {
```

Comment:
Seems unnecessary to have a factory. you can just make the `getLocation` a private function of this class, isn't it?